File: C++/Data Structures/Linked List/Single_Linked_List.cpp
// Title : Single Linked List Implementation using C++
// Author : Ansh Walia <anshwalia@outlook.com>
// Date Created : 21/07/2020

#include <iostream>
#include <stdlib.h>

using namespace std;

// Node Structure
struct Node{
	int data;
	Node * next = NULL;
};

// Single Linked List Class
class Single_Linked_List
{
	private:
		Node * head = NULL;
		Node * currentNode = NULL;
		Node * cursor = NULL;
		int nodeCount = 0;

	public:
		// Linked List Constructor
		Single_Linked_List(Node * head){
			this->head = head;
			this->currentNode = head;
		}

		// Method to Display Linked List
		void display(){
			// If linked list 1st node is NULL
			if(this->head->next == NULL){
				cout << "\nLinked List Empty!" << endl;
			}
			else{
				// Placing cursor at 1st Node
				this->cursor = this->head->next;

				cout << "\nLinked List : (Head)-";

				while(1){
					if(this->cursor->next == NULL){
						cout << "[" << this->cursor->data << "]-(Tail)" << endl;
						break;
					}
					else{
						cout << "[" << this->cursor->data << "]->";
						this->cursor = this->cursor->next;
					}
				}
			}

		}

		// Method to print node count
		void printNodeCount(){
			cout << "\nNode Count : " << this->nodeCount << endl;
		}

		// Method to append node to linked list
		void append(int data){

			// Creating new node and allocating memory dynamically
			Node * newNode = (Node*)malloc(sizeof(Node));

			// Assigning Data to new node
			newNode->data = data;

			// Linking current node to new node
			this->currentNode->next = newNode;

			// Updating current node
			this->currentNode = newNode;

			// Incrementing Node Count
			this->nodeCount += 1;

			cout << "\nOperation : Append Node [ Success ]" << endl;
		}

		// Method to insert node at/after specific value
		void insert(int data,int value){

			if(this->head->next == NULL){
				cout << "\nLinked List Empty!" << endl;
			}
			else{
				// Placing cursor at 1st node
				this->cursor = this->head->next;

				while(1){
					if(this->cursor->data == value){
						// Storing match node in temp variable
						Node * matchNode = this->cursor;

						// Creating New Node
						Node * newNode = (Node*)malloc(sizeof(Node));

						// Assigning Data to new node
						newNode->data = data;

						// Linking new node to next node
						newNode->next = matchNode->next;

						// Linking new to match node
						matchNode->next = newNode;

						// Incrementing Node Count
						this->nodeCount += 1;

						cout << "\nOperation : Insertion [ Success ]" << endl;

						break;
					}
					else{
						if(this->cursor->next == NULL){
							cout << "\nOperation : Insertion [ Failed ]" << endl;
							cout << "No Node Match Found!" << endl;
							break;
						}
						else{
							this->cursor = this->cursor->next;
						}
					}
				}

			}
		}

		// Method to delete node with matching value
		void deleteNode(int value){
			
			if(this->head->next == NULL){
				cout <<  "\nLinked List Empty!" << endl;
			}
			else{
				// Placing Cursor at 1st node
				this->cursor = this->head;

				while(1){
					if(this->cursor->next->data == value){
						// Storing node address of node after target node
						Node * temp = this->cursor->next->next;

						// Freeing memory allocated to unlinked/deleted node
						free(this->cursor->next);

						// Linking node after deleted node to current node
						this->cursor->next = temp;

						// Decrementing Node Count
						this->nodeCount -= 1;

						cout << "\nOperation : Deletion [ Success ]" << endl;

						break;
					}
					else{
						if(this->cursor->next == NULL){
							cout << "\nOperation : Deletion [ Failed ]" << endl;
							cout << "No matching node found!";
							break;
						}
						else{
							this->cursor = this->cursor->next;
						}
					}
				}
					
			}

		}

		// Method to search matching node value
		void search(int value){

			if(this->head->next == NULL){
				cout << "\n[ Linked List Empty ]" << endl;
			}
			else{
				// Placing cursor at 1st node
				this->cursor = this->head->next;

				// Current Node Number
				int nodeNumber = 0;

				while(1){
					// If cursor node value is equal to search value
					if(this->cursor->data == value){
						cout << "\nNode with value [" << value << "], found at Node No. " << nodeNumber << endl;
						break;
					}
					else{
						if(this->cursor->next == NULL){
							cout << "\nNo Node with matching value found!" << endl;
							break;
						}
						else{
							this->cursor = this->cursor->next;
							nodeNumber += 1; 
						}
					}
				}

			}
		}
	
};

int main(){
	// Looping Condition
	bool looping = true;

	// User Option
	int option;

	// Data Input
	int data;

	// Value to find
	int find;

	// Linked List Head
	Node head;

	// Creating New Linked List Object
	Single_Linked_List * sll = new Single_Linked_List(&head);

	while(looping){

		// Linked List Operations Menu
		cout << endl;
		cout << "*****************" << endl;
		cout << "*  LINKED LIST  *" << endl;
		cout << "*****************" << endl;
		cout << "* 1. Append     *" << endl;
		cout << "* 2. Insert     *" << endl;
		cout << "* 3. Delete     *" << endl;
		cout << "* 4. Search     *" << endl;
		cout << "* 5. Display    *" << endl;
		cout << "* 6. Node Count *" << endl;
		cout << "* 0. Exit       *" << endl;
		cout << "*****************" << endl;
		cout << endl;
		cout << "Option : ";

		// User Option Input
		cin >> option;

		// Linux System Call for clearing screen
		system("clear");

		switch(option){

			// Single Linked List [ Operation ] : Node Append
			case 1:
				cout << "\nEnter Data : ";
				cin >> data;
				sll->append(data);
			break;

			// Single Linked List [ Operation ] : Node Insertion
			case 2:
				cout << "\nEnter Data : ";
				cin >> data;
				cout << "\nEnter Value Insert After : ";
				cin >> find;
				sll->insert(data,find);
			break;

			// Single Linked List [ Operation ] : Node Deletion
			case 3:
				cout << "\nEnter Node Value to Delete : ";
				cin >> find;
				sll->deleteNode(find);
			break;

			case 4:
				cout << "\nEnter Node Value to Search : ";
				cin >> find;
				sll->search(find);
			break;

			// Single Linked List 
			case 5:
				sll->display();
			break;

			case 6:
				sll->printNodeCount();
			break;

			case 0:
				cout << "[ Exiting Linked List ]" << endl;
				looping = false;
			break;

			default:
				cout << "Invalid Option! (try again)" << endl;

		}
	}

	return 0;
}

File: Java/Data Structures/Linked List/Single_Linked_List.java
// Title : Single Linked List Implementation Using Java
// Author : Ansh Walia <anshwalia@outlook.com>
// Date Created : 24/07/2020

import java.util.Scanner;

public class Single_Linked_List{
    public static void main(String[] args){
        // Variables
        boolean looping = true;
        int option;
        int data;
        int value;

        // Scanner Object
        Scanner scan = new Scanner(System.in);
        
        // Linked List Object
        LinkedList ll = new LinkedList();

        while(looping){

            // LINKED LIST MENU
            System.out.println("*******************");
            System.out.println("*   LINKED LIST   *");
            System.out.println("*******************");
            System.out.println("* 1. Append       *");
            System.out.println("* 2. Prepend      *");
            System.out.println("* 3. Insert After *");
            System.out.println("* 4. Delete Front *");
            System.out.println("* 5. Delete       *");
            System.out.println("* 6. Delete Back  *");
            System.out.println("* 7. Check Empty  *");
            System.out.println("* 8. Display      *");
            System.out.println("*******************");
            System.out.println("* 0. Exit         *");
            System.out.println("*******************");

            System.out.print("\nOption : ");
            option = scan.nextInt();

            switch(option){

                case 1:
                    System.out.print("\nEnter Data : ");
                    data = scan.nextInt();
                    ll.append(data);
                break;

                case 2:
                    System.out.print("\nEnter Data : ");
                    data = scan.nextInt();
                    ll.prepend(data);
                break;

                case 3:
                    System.out.print("\nEnter Node Value : ");
                    value = scan.nextInt();
                    System.out.print("\nEnter Data : ");
                    data = scan.nextInt();
                    ll.insertAfter(data, value);
                break;

                case 4:
                    ll.deleteFront();
                break;

                case 5:
                    System.out.print("\nEnter Node Value : ");
                    value = scan.nextInt();
                    ll.deleteNode(value);
                break;

                case 6:
                    ll.deleteBack();
                break;

                case 7:
                    if(ll.isEmpty()){
                        System.out.println("\n[ LINKED LIST EMPTY ]\n");
                    }
                    else{
                        System.out.println("\n[ LINKED LIST NOT EMPTY ]\n");
                    }
                break;

                case 8:
                    ll.display();
                break;

                case 0:
                    System.out.println("\n[ EXITING ]\n");
                    looping = false;
                break;
            }
        }

        scan.close();
    }
}

// Node Class
class Node{
    public int data;
    public Node next;

    // Constructor
    Node(int data){
        this.data = data;
        this.next = null;
    }
}

// Linked List
class LinkedList{
    // Properties
    Node head;
    Node tail;
    Node cursor;

    // Node Count
    int nodeCount;

    // Constructor
    LinkedList(){
        this.head = null;
        this.tail = null;
        this.cursor = null;
        this.nodeCount = 0;
    }

    // Method to check if linked list is empty
    public boolean isEmpty(){
        if(this.head == null){
            return true;
        }
        else{
            return false;
        }
    }

    // Method to insert data in Linked List
    public void prepend(int data){
        // Creating New Node
        Node newNode = new Node(data);

        if(this.isEmpty()){
            // Making Head => New Node
            this.head = newNode;
            // Making Tail => New Node
            this.tail = newNode;
        }
        else{
            // Linking New Node to Head's Next Node
            newNode.next = this.head;
            // Making Head's Next Node => New Node
            this.head = newNode;
        }

        // Incrementing Node Count
        this.nodeCount += 1;

        System.out.println("\nOperation : Insertion [ Success ]\n");
    }

    // Method to add node at the end
    public void append(int data){
        // Creating New Node
        Node newNode = new Node(data);

        if(this.isEmpty()){
            // Making Head => New Node
            this.head = newNode;
            // Making Tail => New Node
            this.tail = newNode;
        }
        else{
            // Linking New Node to Tail's Next Node
            tail.next = newNode;
            // Making Head's Next Node => New Node
            tail = newNode;
        }

        // Incrementing Node Count
        this.nodeCount += 1;

        System.out.println("\nOperation : Append [ Success ]\n");
    }

    // Method to insert node after a node with specific value
    public void insertAfter(int data,int value){
        if(this.isEmpty()){
            System.out.println("\n[ LINKED LIST EMPTY ]\n");
        }
        else{
            this.cursor = this.head;

            while(true){
                if(this.cursor.data == value){
                    // Creating New Node
                    Node newNode = new Node(data);

                    // Linking New Node
                    if(this.cursor == this.tail){
                        this.cursor.next = newNode;
                        this.tail = newNode;
                    }
                    else{
                        Node temp = this.cursor.next;
                        this.cursor.next = newNode;
                        newNode.next = temp;
                    }

                    // Incrementing Node Count
                    this.nodeCount += 1;

                    System.out.println("\nOperation : Insert After [ Success ]\n");

                    break;
                }
                else{
                    if(this.cursor.next == null){
                        System.out.println("\nOperation : Insert After [ Failed ]");
                        System.out.println("No Match Found!\n");
                        break;
                    }
                    else{
                        this.cursor = this.cursor.next;
                    }
                }
            }
        }
    }

    // Method to delete node at front of linked list
    public void deleteFront(){
        if(this.isEmpty()){
            System.out.println("\n[ LINKED LIST EMPTY ]\n");
        }
        else{
            if(this.head == this.tail){
                this.head = null;
                this.tail = null;
            }
            else{
                this.head = this.head.next;
            }
            System.out.println("\nOperation : Delete Back [ Success ]\n");
        }
    }

    // Method to delete a node with specific value
    public void deleteNode(int value){
        if(this.isEmpty()){
            System.out.println("\n[ LINKED LIST EMPTY ]\n");
        }
        else{
            this.cursor = this.head;

            if(this.head == this.tail){
                if(this.head.data == value){
                    this.head = null;
                    this.tail = null;
                    System.out.println("\nOperation : Delete Node [ Success ]\n");
                }
                else{
                    System.out.println("\nOperation : Delete Node [ Failed ]");
                    System.out.println("No Match Found!\n");
                }
            }
            else{
                if(this.head.data == value){
                    this.head = this.head.next;
                    System.out.println("\nOperation : Delete Node [ Success ]\n");
                }
                else{
                    while(true){
                        if(this.cursor.next.data == value){
                            Node temp = this.cursor.next.next;
                            this.cursor.next = temp;
                            System.out.println("\nOperation : Delete Node [ Success ]\n");
                            break;
                        }
                        else{
                            if(this.cursor.next == null){
                                System.out.println("\nOperation : Delete Node [ Failed ]");
                                System.out.println("No Match Found!\n");
                                break;
                            }
                            else{
                                this.cursor = this.cursor.next;
                            }
                        }
                    }
                }
            }
        }
    }

    // Method to delete node at end of linked list
    public void deleteBack(){
        if(this.isEmpty()){
            System.out.println("\n[ LINKED LIST EMPTY ]\n");
        }
        else{
            if(this.head == this.tail){
                this.head = null;
                this.tail = null;
            }
            else{
                this.cursor = this.head;
                while(true){
                    if(this.cursor.next == this.tail){
                        this.cursor.next = null;
                        this.tail = this.cursor;
                        break;
                    }
                    else{
                        this.cursor = this.cursor.next;
                    }
                }
            }
            System.out.println("\nOperation : Delete Back [ Success ]\n");
        }
    }

    // Method to display complete Linked List
    public void display(){
        if(this.isEmpty()){
            System.out.println("\n[ LINKED LIST EMPTY ]\n");
        }
        else{
            // Making Cursor => Head
            this.cursor = this.head;
            System.out.print("\nLinked List : (Head)");
            while(this.cursor != null){
                System.out.print("[" + this.cursor.data + "]->");
                this.cursor = this.cursor.next;
            }
            System.out.println("\n");
        }
    }
}